include/bsp/xi_bsp_io_fs.h

```diff
@@ -32,48 +32,131 @@
 extern "C" {
 #endif
 
-typedef intptr_t xi_fs_resource_handle_t;
+typedef intptr_t xi_bsp_io_fs_resource_handle_t;
 
-#define XI_FS_INVALID_RESOURCE_HANDLE -1
-#define xi_fs_init_resource_handle() XI_FS_INVALID_RESOURCE_HANDLE
+#define XI_BSP_IO_FS_INVALID_RESOURCE_HANDLE -1
+#define xi_bsp_io_fs_init_resource_handle() XI_BSP_IO_FS_INVALID_RESOURCE_HANDLE
 
 /**
- * @enum xi_fs_resource_type_t
+ * @typedef xi_bsp_io_net_state_e
+ * @brief Return value of the BSP NET API functions.
+ *
+ * The implementation reports internal status to Xively Client through these values.
+ */
+ typedef enum xi_bsp_io_fs_state_e {
+    /** operation finished successfully */
+    XI_BSP_IO_FS_STATE_OK = 0,
+    /** operation failed on generic error */
+    XI_BSP_IO_FS_ERROR = 1,
+    /** invalid parameter passed to function */
+    XI_BSP_IO_FS_INVALID_PARAMETER = 2,
+    /** resource/file is not available */
+    XI_BSP_IO_FS_RESOURCE_NOT_AVAILABLE = 3,
+    /** out of memory error */
+    XI_BSP_IO_FS_OUT_OF_MEMORY = 4,
+    /** function not implemented on target platform */
+    XI_BSP_IO_FS_NOT_IMPLEMENTED = 5,
+    /** error opening file resource **/
+    XI_BSP_IO_FS_OPEN_ERROR = 6,
+    /** error that file open could only open read only */
+    XI_BSP_IO_FS_OPEN_READ_ONLY = 7,
+    /** error reported when file cannot be removed */
+    XI_BSP_IO_FS_REMOVE_ERROR = 8,
+    /** error when attempting to write file data */
+    XI_BSP_IO_FS_WRITE_ERROR = 9,
+    /** error reported when file cannot be read */
+    XI_BSP_IO_FS_READ_ERROR = 10,
+    /** error reported when file cannot be closed */
+    XI_BSP_IO_FS_CLOSE_ERROR = 11,
+} xi_bsp_io_fs_state_t;
+
+/* helper function that translates the errors to the xi_bsp_io_fs_state_t values */
+/* TODO: remove the need to have this inline, as the function has a substantial jump
+ * table. */
+static inline xi_state_t xi_fs_bsp_io_fs_2_xi_state( xi_bsp_io_fs_state_t bsp_state_value )
+{
+    xi_state_t ret = XI_STATE_OK;
+
+    switch ( bsp_state_value )
```

Comment:
what is the purpose of the new, separated BSP enum introduction? I thought the purpose is to hide the xively_error.h values. But they are not hidden. Other thought was to prevent the customer implemented BSP IO FS layer to return non IO FS related errors e.g. XI_SOCKET_GETHOSTBYNAME_ERROR or something total non-related. But as I see the BSP is responsible to convert IO_FS_ERRORS to libxively_ERROS.
I would do one of the following:
1. let BSP IO FS use xively_error.h errors
2. move the conversion (implementation of this function) to xi_fs_posix.c

Reply:
Yeah, I see your point but I did it this way for a few of reasons:


1. I used the Networking BSP and TLS BSP as a standard method of practice, and each of them its own BSP Error codes.  I was trying to be consistent.

2. It creates a very succinct vocabulary for filling out a particular BSP.  As an engineer I know exactly what’s expected due to the limited number of enumerations.  This is important for easing the development of new systems.  You and I know might know what sort of error codes the invoking code expects back, what are good errors to return, etc, but are new engineers coming to the code base so well acquainted with what to do?

2a. If we use the main error codes, which is an extensive list, it gets confusing for developers.  I’ve been in this scenario before where there were global error enumerations, butI don’t know which were valid errors to return for a given subsystem and I didn’t know if I could cause problems returning wrong error codes.  I also took a lot of time guessing what I SHOULD be returning, etc, where it woul be much nicer to have that list be a small and focused as possible.

Mostly for 1… but I do see a lot of value in 2…. Too.